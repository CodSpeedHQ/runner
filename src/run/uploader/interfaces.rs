use serde::{Deserialize, Serialize};

use crate::run::{
    check_system::SystemInfo,
    ci_provider::interfaces::{CIProviderMetadata, RepositoryProvider},
    instruments::InstrumentName,
    runner::ExecutorName,
};

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UploadMetadata {
    pub repository_provider: RepositoryProvider,
    pub version: Option<u32>,
    pub tokenless: bool,
    pub profile_md5: String,
    pub runner: Runner,
    pub platform: String,
    pub commit_hash: String,
    #[serde(flatten)]
    pub ci_provider_metadata: CIProviderMetadata,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Runner {
    pub name: String,
    pub version: String,
    pub instruments: Vec<InstrumentName>,
    pub executor: ExecutorName,
    #[serde(flatten)]
    pub system_info: SystemInfo,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UploadData {
    pub status: String,
    pub upload_url: String,
    pub run_id: String,
}

#[derive(Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UploadError {
    pub error: String,
}
