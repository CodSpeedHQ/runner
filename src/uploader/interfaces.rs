use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UploadMetadata {
    pub version: Option<u32>,
    pub tokenless: bool,
    #[serde(rename = "ref")]
    pub ref_: String,
    pub head_ref: Option<String>,
    pub base_ref: Option<String>,
    pub owner: String,
    pub repository: String,
    pub commit_hash: String,
    pub event: String,
    pub profile_md5: String,
    pub gh_data: GhData,
    pub runner: Runner,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct GhData {
    pub run_id: u32,
    pub job: String,
    pub sender: Option<Sender>,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Sender {
    pub id: u32,
    pub login: String,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Runner {
    pub name: String,
    pub version: String,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PostResponse {
    pub status: String,
    pub upload_url: String,
    pub run_id: String,
}
