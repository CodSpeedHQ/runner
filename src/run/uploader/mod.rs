mod interfaces;
mod upload;
mod upload_metadata;

pub use interfaces::*;
pub use upload::upload;
