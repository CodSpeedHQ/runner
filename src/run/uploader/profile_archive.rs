use base64::{Engine, engine::general_purpose};

use crate::prelude::*;
use std::path::PathBuf;

#[derive(Debug)]
pub struct ProfileArchive {
    pub hash: String,
    pub content: ProfileArchiveContent,
}

#[derive(Debug)]
pub enum ProfileArchiveContent {
    CompressedInMemory { data: Vec<u8> },
    UncompressedOnDisk { path: PathBuf },
}

impl ProfileArchive {
    pub fn new_compressed_in_memory(data: Vec<u8>) -> Self {
        let hash = general_purpose::STANDARD.encode(md5::compute(&data).0);
        ProfileArchive {
            hash,
            content: ProfileArchiveContent::CompressedInMemory { data },
        }
    }

    pub fn new_uncompressed_on_disk(path: PathBuf) -> Result<Self> {
        let metadata = std::fs::metadata(&path)?;
        if !metadata.is_file() {
            return Err(anyhow!("The provided path is not a file"));
        }
        let mut file = std::fs::File::open(&path)?;
        let mut buffer = Vec::new();
        use std::io::Read;
        file.read_to_end(&mut buffer)?;
        let hash = general_purpose::STANDARD.encode(md5::compute(&buffer).0);
        Ok(ProfileArchive {
            hash,
            content: ProfileArchiveContent::UncompressedOnDisk { path },
        })
    }
}

impl ProfileArchiveContent {
    pub async fn size(&self) -> Result<u64> {
        match &self {
            ProfileArchiveContent::CompressedInMemory { data } => Ok(data.len() as u64),
            ProfileArchiveContent::UncompressedOnDisk { path } => {
                let metadata = tokio::fs::metadata(path).await?;
                Ok(metadata.len())
            }
        }
    }

    pub fn encoding(&self) -> Option<String> {
        match self {
            ProfileArchiveContent::CompressedInMemory { .. } => Some("gzip".to_string()),
            _ => None,
        }
    }
}

impl Drop for ProfileArchiveContent {
    fn drop(&mut self) {
        if let ProfileArchiveContent::UncompressedOnDisk { path } = self {
            if path.exists() {
                let _ = std::fs::remove_file(path);
            }
        }
    }
}
