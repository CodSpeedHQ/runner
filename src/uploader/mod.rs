mod interfaces;
mod upload;

pub use upload::upload;
