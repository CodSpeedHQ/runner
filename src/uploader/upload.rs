use async_compression::tokio::write::GzipEncoder;
use base64::{engine::general_purpose, Engine as _};
use lazy_static::lazy_static;
use tokio::io::AsyncWriteExt;
use tokio_tar::Builder;

use crate::{config::Config, prelude::*, runner::RunData};
use reqwest::ClientBuilder;
use reqwest_middleware::{ClientBuilder as ClientWithMiddlewareBuilder, ClientWithMiddleware};
use reqwest_retry::{policies::ExponentialBackoff, RetryTransientMiddleware};
use serde_json::json;

use super::{
    ci_provider::CIProvider,
    github_actions_provider::GitHubActionsProvider,
    interfaces::{UploadData, UploadMetadata},
};

fn get_provider(config: &Config) -> Result<impl CIProvider> {
    if GitHubActionsProvider::detect() {
        let provider = GitHubActionsProvider::try_from(config)?;
        return Ok(provider);
    }

    bail!("No CI provider detected")
}

/// Create a tar.gz archive buffer of the profile folder and return its md5 hash encoded in base64
async fn get_profile_archive_buffer(run_data: &RunData) -> Result<(Vec<u8>, String)> {
    let enc = GzipEncoder::new(Vec::new());
    let mut tar = Builder::new(enc);
    tar.append_dir_all(".", run_data.profile_folder.clone())
        .await?;
    let mut gzip_encoder = tar.into_inner().await?;
    gzip_encoder.shutdown().await?;

    let archive_buffer = gzip_encoder.into_inner();
    let archive_digest = md5::compute(archive_buffer.as_slice());
    let archive_hash = general_purpose::STANDARD.encode(archive_digest.0);

    Ok((archive_buffer, archive_hash))
}

const UPLOAD_RETRY_COUNT: u32 = 3;

lazy_static! {
    static ref UPLOAD_REQUEST_CLIENT: ClientWithMiddleware = ClientWithMiddlewareBuilder::new(
        ClientBuilder::new()
            .user_agent("codspeed-runner")
            .build()
            .unwrap()
    )
    .with(RetryTransientMiddleware::new_with_policy(
        ExponentialBackoff::builder().build_with_max_retries(UPLOAD_RETRY_COUNT)
    ))
    .build();
}

async fn retrieve_upload_data(
    config: &Config,
    upload_metadata: &UploadMetadata,
) -> Result<UploadData> {
    let mut upload_request = UPLOAD_REQUEST_CLIENT
        .post(config.upload_url.clone())
        .json(&upload_metadata);
    if !upload_metadata.tokenless {
        upload_request = upload_request.header("Authorization", config.token.clone().unwrap());
    }

    let response = upload_request.send().await?.json::<UploadData>().await?;

    Ok(response)
}

async fn upload_archive_buffer(
    upload_data: &UploadData,
    archive_buffer: Vec<u8>,
    archive_hash: &String,
) -> Result<()> {
    UPLOAD_REQUEST_CLIENT
        .put(upload_data.upload_url.clone())
        .header("Content-Type", "application/gzip")
        .header("Content-Length", archive_buffer.len())
        .header("Content-MD5", archive_hash)
        .body(archive_buffer)
        .send()
        .await?;

    Ok(())
}

pub async fn upload(config: &Config, run_data: &RunData) -> Result<()> {
    let (archive_buffer, archive_hash) = get_profile_archive_buffer(run_data).await?;

    let provider = get_provider(config)?;
    debug!("CI provider detected: {:#?}", provider.get_provider_name());

    let upload_metadata = provider.get_upload_metadata(config, &archive_hash)?;
    debug!("Upload metadata: {:#?}", upload_metadata);
    if upload_metadata.tokenless {
        let hash = sha256::digest(json!(upload_metadata).to_string());
        info!("CodSpeed Run Hash: {}", hash);
    }

    info!("Preparing upload...");
    let upload_data = retrieve_upload_data(config, &upload_metadata).await?;
    debug!("runId: {}", upload_data.run_id);

    info!("Uploading profile data...");
    debug!("Uploading {} bytes...", archive_buffer.len());
    upload_archive_buffer(&upload_data, archive_buffer, &archive_hash).await?;
    info!("Results uploaded.");

    Ok(())
}

#[cfg(test)]
mod tests {
    use temp_env::async_with_vars;
    use url::Url;

    use super::*;
    use crate::runner::RunData;
    use std::path::PathBuf;

    // TODO: remove the ignore when implementing network mocking
    #[ignore]
    #[tokio::test]
    async fn test_upload() {
        let config = Config {
            command: "pytest tests/ --codspeed".into(),
            upload_url: Url::parse("change me").unwrap(),
            skip_setup: false,
            skip_upload: false,
            token: Some("change me".into()),
            working_directory: None,
        };
        let run_data = RunData {
            profile_folder: PathBuf::from(format!(
                "{}/src/uploader/samples/adrien-python-test",
                env!("CARGO_MANIFEST_DIR")
            )),
        };
        async_with_vars(
            [
                ("GITHUB_ACTIONS", Some("true")),
                ("GITHUB_ACTOR_ID", Some("19605940")),
                ("GITHUB_ACTOR", Some("adriencaccia")),
                ("GITHUB_BASE_REF", Some("main")),
                ("GITHUB_EVENT_NAME", Some("pull_request")),
                (
                    "GITHUB_EVENT_PATH",
                    Some(
                        format!(
                            "{}/src/uploader/samples/pr-event.json",
                            env!("CARGO_MANIFEST_DIR")
                        )
                        .as_str(),
                    ),
                ),
                ("GITHUB_HEAD_REF", Some("feat/codspeed-runner")),
                ("GITHUB_JOB", Some("log-env")),
                ("GITHUB_REF", Some("refs/pull/22/merge")),
                ("GITHUB_REPOSITORY", Some("my-org/adrien-python-test")),
                ("GITHUB_RUN_ID", Some("6957110437")),
                (
                    "GITHUB_SHA",
                    Some("5bd77cb0da72bef094893ed45fb793ff16ecfbe3"),
                ),
                ("VERSION", Some("0.1.0")),
            ],
            async {
                upload(&config, &run_data).await.unwrap();
            },
        )
        .await;
    }
}
