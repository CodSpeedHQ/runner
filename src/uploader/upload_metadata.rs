use serde_json::json;

use super::UploadMetadata;

impl UploadMetadata {
    pub fn get_hash(&self) -> String {
        let upload_metadata_string = json!(&self).to_string();
        sha256::digest(upload_metadata_string)
    }
}

#[cfg(test)]
mod tests {
    use crate::uploader::{GhData, RunEvent, Runner, Sender, UploadMetadata};

    #[test]
    fn test_get_metadata_hash() {
        let upload_metadata = UploadMetadata {
            version: Some(1),
            tokenless: true,
            ref_: "refs/pull/29/merge".into(),
            head_ref: Some("chore/native-action-runner".into()),
            base_ref: Some("main".into()),
            owner: "CodSpeedHQ".into(),
            repository: "codspeed-node".into(),
            commit_hash: "ea4005444338762d85163c8e8787387e2ba97fb6".into(),
            event: RunEvent::PullRequest,
            profile_md5: "jp/k05RKuqP3ERQuIIvx4Q==".into(),
            gh_data: Some(GhData {
                run_id: 7044765741,
                job: "codspeed".into(),
                sender: Some(Sender {
                    id: 19605940,
                    login: "adriencaccia".into(),
                }),
            }),
            runner: Runner {
                name: "codspeed-runner".into(),
                version: "2.0.1-beta.6".into(),
            },
            platform: "github-actions".into(),
            repository_root_path: "/home/runner/work/codspeed-node/codspeed-node/".into(),
        };

        let hash = upload_metadata.get_hash();
        assert_eq!(
            hash,
            "ca9a71d06e61517fabd2784eb690b1cb8a9736f93122d07f20abb8e5356761ca"
        )
    }
}
