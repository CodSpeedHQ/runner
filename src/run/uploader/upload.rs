use crate::run::{
    check_system::SystemInfo, ci_provider::CIProvider, config::Config, runner::RunData,
    uploader::UploadError,
};
use crate::{prelude::*, request_client::REQUEST_CLIENT};
use async_compression::tokio::write::GzipEncoder;
use base64::{engine::general_purpose, Engine as _};
use console::style;
use tokio::io::AsyncWriteExt;
use tokio_tar::Builder;

use super::interfaces::{UploadData, UploadMetadata};

/// Create a tar.gz archive buffer of the profile folder and return its md5 hash encoded in base64
async fn get_profile_archive_buffer(run_data: &RunData) -> Result<(Vec<u8>, String)> {
    let enc = GzipEncoder::new(Vec::new());
    let mut tar = Builder::new(enc);
    tar.append_dir_all(".", run_data.profile_folder.clone())
        .await?;
    let mut gzip_encoder = tar.into_inner().await?;
    gzip_encoder.shutdown().await?;

    let archive_buffer = gzip_encoder.into_inner();
    let archive_digest = md5::compute(archive_buffer.as_slice());
    let archive_hash = general_purpose::STANDARD.encode(archive_digest.0);

    Ok((archive_buffer, archive_hash))
}

async fn retrieve_upload_data(
    config: &Config,
    upload_metadata: &UploadMetadata,
) -> Result<UploadData> {
    let mut upload_request = REQUEST_CLIENT
        .post(config.upload_url.clone())
        .json(&upload_metadata);
    if !upload_metadata.tokenless {
        upload_request = upload_request.header("Authorization", config.token.clone().unwrap());
    }

    let response = upload_request.send().await;

    match response {
        Ok(response) => {
            if response.status().is_client_error() {
                let status = response.status();
                let text = response.text().await?;
                let error_message = serde_json::from_str::<UploadError>(&text)
                    .map(|body| body.error)
                    .unwrap_or(text);
                bail!(
                    "Failed to retrieve upload data: {}\n  -> {} {}",
                    status,
                    style("Reason:").bold(),
                    // !!! we have to manually apply `.red()` this because the `.red()` from the logger is not applied to everything that follows the first `style()` call, for some reasons 🤷‍♂️🤷‍♂️🤷‍♂️
                    style(error_message).red()
                );
            }

            Ok(response.json().await?)
        }
        Err(err) => Err(err.into()),
    }
}

async fn upload_archive_buffer(
    upload_data: &UploadData,
    archive_buffer: Vec<u8>,
    archive_hash: &String,
) -> Result<()> {
    REQUEST_CLIENT
        .put(upload_data.upload_url.clone())
        .header("Content-Type", "application/gzip")
        .header("Content-Length", archive_buffer.len())
        .header("Content-MD5", archive_hash)
        .body(archive_buffer)
        .send()
        .await?;

    Ok(())
}

pub struct UploadResult {
    pub run_id: String,
}

#[allow(clippy::borrowed_box)]
pub async fn upload(
    config: &Config,
    system_info: &SystemInfo,
    provider: &Box<dyn CIProvider>,
    run_data: &RunData,
) -> Result<UploadResult> {
    let (archive_buffer, archive_hash) = get_profile_archive_buffer(run_data).await?;

    debug!("CI provider detected: {:#?}", provider.get_provider_name());

    let upload_metadata = provider.get_upload_metadata(config, system_info, &archive_hash)?;
    debug!("Upload metadata: {:#?}", upload_metadata);
    info!(
        "Repository {} detected\n",
        style(format!(
            "{}/{}",
            upload_metadata.provider_metadata.owner, upload_metadata.provider_metadata.repository
        ))
        .bold(),
    );
    if upload_metadata.tokenless {
        let hash = upload_metadata.get_hash();
        info!("CodSpeed Run Hash: \"{}\"", hash);
    }

    info!("Preparing upload...");
    let upload_data = retrieve_upload_data(config, &upload_metadata).await?;
    debug!("runId: {}", upload_data.run_id);

    info!("Uploading performance data...");
    debug!("Uploading {} bytes...", archive_buffer.len());
    upload_archive_buffer(&upload_data, archive_buffer, &archive_hash).await?;
    info!("Performance data uploaded");

    Ok(UploadResult {
        run_id: upload_data.run_id,
    })
}

#[cfg(test)]
mod tests {
    use temp_env::async_with_vars;
    use url::Url;

    use super::*;
    use std::path::PathBuf;

    // TODO: remove the ignore when implementing network mocking
    #[ignore]
    #[tokio::test]
    async fn test_upload() {
        let config = Config {
            command: "pytest tests/ --codspeed".into(),
            upload_url: Url::parse("change me").unwrap(),
            token: Some("change me".into()),
            ..Config::test()
        };
        let run_data = RunData {
            profile_folder: PathBuf::from(format!(
                "{}/src/uploader/samples/adrien-python-test",
                env!("CARGO_MANIFEST_DIR")
            )),
        };
        let system_info = SystemInfo::test();
        async_with_vars(
            [
                ("GITHUB_ACTIONS", Some("true")),
                ("GITHUB_ACTOR_ID", Some("19605940")),
                ("GITHUB_ACTOR", Some("adriencaccia")),
                ("GITHUB_BASE_REF", Some("main")),
                ("GITHUB_EVENT_NAME", Some("pull_request")),
                (
                    "GITHUB_EVENT_PATH",
                    Some(
                        format!(
                            "{}/src/uploader/samples/pr-event.json",
                            env!("CARGO_MANIFEST_DIR")
                        )
                        .as_str(),
                    ),
                ),
                ("GITHUB_HEAD_REF", Some("feat/codspeed-runner")),
                ("GITHUB_JOB", Some("log-env")),
                ("GITHUB_REF", Some("refs/pull/22/merge")),
                ("GITHUB_REPOSITORY", Some("my-org/adrien-python-test")),
                ("GITHUB_RUN_ID", Some("6957110437")),
                (
                    "GITHUB_SHA",
                    Some("5bd77cb0da72bef094893ed45fb793ff16ecfbe3"),
                ),
                ("VERSION", Some("0.1.0")),
            ],
            async {
                let provider = crate::run::ci_provider::get_provider(&config).unwrap();
                upload(&config, &system_info, &provider, &run_data)
                    .await
                    .unwrap();
            },
        )
        .await;
    }
}
