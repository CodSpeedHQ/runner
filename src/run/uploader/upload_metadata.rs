use serde_json::json;

use super::UploadMetadata;

impl UploadMetadata {
    pub fn get_hash(&self) -> String {
        let upload_metadata_string = json!(&self).to_string();
        sha256::digest(upload_metadata_string)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;

    use crate::run::{
        check_system::SystemInfo,
        ci_provider::interfaces::{GhData, ProviderMetadata, RunEvent, Sender},
        instruments::InstrumentName,
        runner::ExecutorName,
        uploader::{Runner, UploadMetadata},
    };

    #[test]
    fn test_get_metadata_hash() {
        let upload_metadata = UploadMetadata {
            version: Some(4),
            tokenless: true,
            profile_md5: "jp/k05RKuqP3ERQuIIvx4Q==".into(),
            runner: Runner {
                name: "codspeed-runner".into(),
                version: "2.1.0".into(),
                instruments: vec![InstrumentName::MongoDB],
                executor: ExecutorName::Valgrind,
                system_info: SystemInfo::test(),
            },
            platform: "github-actions".into(),
            commit_hash: "5bd77cb0da72bef094893ed45fb793ff16ecfbe3".into(),
            provider_metadata: ProviderMetadata {
                ref_: "refs/pull/29/merge".into(),
                head_ref: Some("chore/native-action-runner".into()),
                base_ref: Some("main".into()),
                owner: "CodSpeedHQ".into(),
                repository: "codspeed-node".into(),
                event: RunEvent::PullRequest,
                gh_data: Some(GhData {
                    run_id: "7044765741".into(),
                    job: "codspeed".into(),
                    sender: Some(Sender {
                        id: 19605940,
                        login: "adriencaccia".into(),
                    }),
                }),
                repository_root_path: "/home/runner/work/codspeed-node/codspeed-node/".into(),
            },
        };

        let hash = upload_metadata.get_hash();
        assert_eq!(
            hash,
            "b85b85835e4fc5d222e35fa762353dfcfecbed157ee866323bffa4c11151f66d"
        );
        assert_json_snapshot!(upload_metadata);
    }
}
