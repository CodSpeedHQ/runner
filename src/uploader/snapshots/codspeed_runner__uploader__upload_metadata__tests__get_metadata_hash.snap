---
source: src/uploader/upload_metadata.rs
expression: upload_metadata
---
{
  "version": 2,
  "tokenless": true,
  "profileMd5": "jp/k05RKuqP3ERQuIIvx4Q==",
  "runner": {
    "name": "codspeed-runner",
    "version": "2.1.0",
    "instruments": [
      "MongoDB"
    ]
  },
  "platform": "github-actions",
  "ref": "refs/pull/29/merge",
  "headRef": "chore/native-action-runner",
  "baseRef": "main",
  "owner": "CodSpeedHQ",
  "repository": "codspeed-node",
  "actualCommitHash": "5bd77cb0da72bef094893ed45fb793ff16ecfbe3",
  "commitHash": "ea4005444338762d85163c8e8787387e2ba97fb6",
  "event": "pull_request",
  "ghData": {
    "runId": 7044765741,
    "job": "codspeed",
    "sender": {
      "id": 19605940,
      "login": "adriencaccia"
    }
  },
  "repositoryRootPath": "/home/runner/work/codspeed-node/codspeed-node/"
}
